### ISSUE TITLE
Deprecate 'function' keyword

### ISSUE THREAD

**Issue by messengerbag** [adventuregamestudio/ags#2463]
**Describe the problem**
The `function` keyword is a legacy of ancient AGS scripting. Many of the default functions that are called by the engine use it as their return type. Because it is implemented as an alias of `int` but is used even if the function doesn't actually return anything, it means that AGS script cannot properly distinguish `int` and `void` return types.

It also creates inconsistency, with two/three different ways to write the same function. This arguably creates confusion and makes the AGS scripting language harder to learn.

**Suggested change**
Remove `function` as a keyword. Change all function definitions to use `int` or `void` as appropriate, and make sure the editor uses these when creating event handlers. This also requires an update to the manual and to the templates.

Optionally, update the compiler to require a correct return type (do not allow a bare `return;` if the return type is declared as `int`, or a `return 1;` if the return type is declared as `void`).

Optionally, retain `function` as a keyword (an alias for `int`) **if** backwards script compatibility is set. (Note that if the other optional part is implemented, this would still require users to change all functions that need to have `void` as their return type, unless that is also disabled if the script compatibility option is set.)

**Alternative suggestion**
Alternatively, don't remove `function` as a keyword, but change all the built-in functions to either `int` or `void` as appropriate (including in the manual and templates), so that `function` never appears unless you specifically add it.

**Comment by AlanDrake** [comment]
I guess I'll go check the manual for old examples with function.

**Comment by ericoporto** [comment]
About ags3 , check https://github.com/adventuregamestudio/ags/issues/1331#issuecomment-877709952

I think void in ags3 is actually int for compatibility because some old games returned something in a `void` function and it was kept to avoid breaking old games.

**Comment by messengerbag** [comment]
Yes, I thought so. That's part of why I think AGS 4.0 is a good opportunity to fix it.

**Comment by ivan-mogilko** [comment]
> I think void in ags3 is actually int for compatibility because some old games returned something in a void function and it was kept to avoid breaking old games.

To clarify, today `void` is a valid keyword in compiler and not an `int` alias, but the engine "secretly" still returns a integer '0' from the functions of void type. This is done because old API was not declared as `void` but as `function`, even though it did not return anything meaningful, and unfortunately some users have used this return value by mistake (it was always 0).

**Comment by fernewelten** [comment]
> I think void in ags3 is actually int for compatibility

The old compiler returns `0` in the AX register for `void` functions IIRC.

The new compiler (currently AGS4 only) doesn't: 
- When a function is `void` then the compiler doesn't load the AX register with anything when the function returns. 
- OTOH, when a function is `int` and code falls through to the end, then the function will load  `0` into the AX register.

**Comment by fernewelten** [comment]
The header file for AGS code has `#define function int`. In my opinion, that's a kludge in and of itself. It allows absurd code such as: 

```
function test(function param)
{
    function i = 5, j = 5;
    for (function k = 0; k < j; k++)
        i += k;
}
```

**Comment by ivan-mogilko** [comment]
I think there are two reasonable paths:

If we want to keep "function" as a valid keyword, then it must be implemented in compiler as a keyword, with the rule to be only used in function declaration/definition.

If we want to stop using "function", then:
1. In standard api header this macro declaration may be hidden under "#ifdef SCRIPT_COMPAT_v399" (for instance). This will disable it in ags4, but still let users to enable if they switch compatibility level.
2. Replace "function" with "void" or "int" as appropriate for all the standard api declarations.
3. Replace these in the ags4 templates.
4. Replace these in the manual (for ags4 only? idk).

**Comment by AlanDrake** [comment]
When I see a `function` without extra type definition syntaxes, I expect that it can return any type.
I guess this could be achieved by hunting for return statements and resolving their type during compile, but unless we actually do that it may be better to remove it from ags4.

For 4. and the current manual, I could replace `function` from entries that aren't explicitly generated by the editor. I actually have this change already prepared in case it get greenlighted.

**Comment by ericoporto** [comment]
https://github.com/adventuregamestudio/ags/blob/51b0ba178a0a00fb921dc18ac0b1a40a7919b5a5/Editor/AGS.Editor/Resources/agsdefns.sh#L2713

There are a ton of things in the Character API that should be void but are functions. Curiously the old global API uses void more there. Can these just be made void in ags4 - I mean without the need of switches. I believe this is confusing because it won't trigger a compiler error if theie return value is used.

**Comment by ivan-mogilko** [comment]
> There are a ton of things in the Character API that should be void but are functions. Curiously the old global API uses void more there. Can these just be made void in ags4 - I mean without the need of switches. I believe this is confusing because it won't trigger a compiler error if theie return value is used.

These may be freely changed to void in both ags3 and ags4, as that won't break any old games, and may help to notice mistakes when upgrading an older project too.
